### PR TITLE
Detect unused struct impls pub trait

### LINKED ISSUE #47851: impls should not inhibit the unused_item lint
```rust
#![crate_type="lib"]

enum Foo {
}

impl Clone for Foo {
    fn clone(&self) -> Foo { loop {} }
}
```

This code emits no warnings because `impl Clone for Foo` is considered a "use" of the enum, however, implementation for this enum cannot possibly be used either, as the enum is private and is unused. This is more notable for `derive`d implementations, which also inhibit this lint.

The inherent implementations behave as expected.

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #47851

### CODE REVIEW

File: tests/ui/derives/clone-debug-dead-code-in-the-same-struct.stderr [hunk 1]
```diff
@@ -13,7 +13,7 @@ LL |     field3: (),
 LL |     field4: (),
    |     ^^^^^^
    |
-   = note: `Whatever` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
+   = note: `Whatever` has derived impls for the traits `Debug` and `Debug`, but these are intentionally ignored during dead code analysis
```
Comment (clubby789) [inline tests/ui/derives/clone-debug-dead-code-in-the-same-struct.stderr:16]: This seems incorrect?

**Comment by shepmaster** [comment]
I'll give this a look-through, but I'm definitely not qualified to give it a final sign-off 🙂.

r? rust-lang/compiler

**Comment by klensy** [comment]
Push after r+, need reapprove?

**Comment by pnkfelix** [comment]
@mu001999 it looks like the performance regression flagged here is legitimate, even though it is also minor. I.e., the performance stats indicate that the newly added worklist code is the source of a bit of extra time we are spending here.

Do you have any ideas on ways we might attempt to avoid the extra cost?

If not, that is okay; we just need to keep our eyes on how these things add up, and make sure we don't end up slowing down the compiler too much for lints that may not be paying for themselves.

**Comment by pnkfelix** [comment]
Yeah that sounds like a potential idea. Its not something you need to take on yourself, I just want to try to get the ideas written down while the code is still fresh in our heads.

**Pull request merged**